### PR TITLE
Wire in additional PG Gate APIs and clean up

### PR DESCRIPTION
**Pull request by johnfangAFW**
1) wire in PgGate_GetTableDesc, PgGate_GetColumnInfo, PgGate_SetIsSysCatalogVersionChange, and PgGate_SetCatalogCacheVersion
2)  add foreign key constraint logic
3)  remove unneeded pg_gate apis

### CODE REVIEW

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp [hunk 1]
```diff
@@ -904,52 +787,29 @@ K2PgStatus PgGate_OperatorAppendArg(K2PgExpr op_handle, K2PgExpr arg){
 // Referential Integrity Check Caching.
 // Check if foreign key reference exists in cache.
 bool PgGate_ForeignKeyReferenceExists(K2PgOid table_oid, const char* k2pgctid, int64_t k2pgctid_size) {
-  elog(DEBUG5, "PgGateAPI: PgGate_ForeignKeyReferenceExists %d, %s, %ld", table_oid, k2pgctid, k2pgctid_size);
-  return false;
+    elog(DEBUG5, "PgGateAPI: PgGate_ForeignKeyReferenceExists %d, %s, %ld", table_oid, k2pgctid, k2pgctid_size);
```
Comment (jfunston) [inline src/gausskernel/storage/access/k2/pg_gate_api.cpp]: I don't think its safe to print k2pgctid as a string, maybe just print the pointer value

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp [hunk 2]
```diff
@@ -904,52 +787,29 @@ K2PgStatus PgGate_OperatorAppendArg(K2PgExpr op_handle, K2PgExpr arg){
...
+    return k2pg::pg_session->ForeignKeyReferenceExists(table_oid, std::string(k2pgctid, k2pgctid_size));
 }
 
 // Add an entry to foreign key reference cache.
 K2PgStatus PgGate_CacheForeignKeyReference(K2PgOid table_oid, const char* k2pgctid, int64_t k2pgctid_size){
-  elog(DEBUG5, "PgGateAPI: PgGate_CacheForeignKeyReference %d, %s, %ld", table_oid, k2pgctid, k2pgctid_size);
-  return K2PgStatus::NotSupported;
+    elog(DEBUG5, "PgGateAPI: PgGate_CacheForeignKeyReference %d, %s, %ld", table_oid, k2pgctid, k2pgctid_size);
```
Comment (jfunston) [inline src/gausskernel/storage/access/k2/pg_gate_api.cpp]: Same as above on printing k2pgctid

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp [hunk 3]
```diff
@@ -904,52 +787,29 @@ K2PgStatus PgGate_OperatorAppendArg(K2PgExpr op_handle, K2PgExpr arg){
...
 
 // Delete an entry from foreign key reference cache.
 K2PgStatus PgGate_DeleteFromForeignKeyReferenceCache(K2PgOid table_oid, uint64_t k2pgctid){
-  elog(DEBUG5, "PgGateAPI: PgGate_DeleteFromForeignKeyReferenceCache %d, %lu", table_oid, k2pgctid);
-  return K2PgStatus::NotSupported;
+    elog(DEBUG5, "PgGateAPI: PgGate_DeleteFromForeignKeyReferenceCache %d, %lu", table_oid, k2pgctid);
+    char *value;
+    int64_t bytes;
+    const K2PgTypeEntity *type_entity = pg_gate->FindTypeEntity(k2pg::kPgByteArrayOid);
+    type_entity->datum_to_k2pg(k2pgctid, &value, &bytes);
```
Comment (jfunston) [inline src/gausskernel/storage/access/k2/pg_gate_api.cpp]: I think you should convert the datum like I do in storage.cpp. So something like:
        UntoastedDatum data = UntoastedDatum(k2pgctid);
        bytes = VARSIZE(data.untoasted);  // includes header len VARHDRSZ
        value = (char*)data.untoasted

File: src/gausskernel/storage/access/k2/pg_session.cpp [hunk 4]
```diff
@@ -91,4 +91,33 @@ std::shared_ptr<PgTableDesc> PgSession::LoadTable(const PgObjectId& table_object
...
+}
+
+size_t hash_value(const PgForeignKeyReference& key) {
+  size_t hash = 0;
+  boost::hash_combine(hash, key.table_oid);
+  boost::hash_combine(hash, key.k2pgctid);
+  return hash;
+}
+
+bool PgSession::ForeignKeyReferenceExists(uint32_t table_oid, std::string&& k2pgctid) {
```
Comment (jfunston) [inline src/gausskernel/storage/access/k2/pg_session.cpp:106]: Do you know how this foreign key cache is used? I'm wondering if we need to make this distributed in k2 to work properly with distributed sql instances.

**Review by jfunston — CHANGES_REQUESTED** [review]

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp (hunk 1 above)
Comment (johnfangAFW) [reply src/gausskernel/storage/access/k2/pg_gate_api.cpp]: good point, changed to print pointer address.

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp (hunk 2 above)
Comment (johnfangAFW) [reply src/gausskernel/storage/access/k2/pg_gate_api.cpp]: changed to print pointer address

File: src/gausskernel/storage/access/k2/pg_gate_api.cpp (hunk 3 above)
Comment (johnfangAFW) [reply src/gausskernel/storage/access/k2/pg_gate_api.cpp]: moved UntoastedDatum to k2_util.h and changed as suggested.

File: src/gausskernel/storage/access/k2/pg_session.cpp (hunk 4 above)
Comment (johnfangAFW) [reply src/gausskernel/storage/access/k2/pg_session.cpp:106]: foreign key cache is used for a tuple insert or delete, which is checked by postgres internal logic based on foreign key constraints in table/index schema. That is why the foreign key cache is only pg_session based and there is no need to distribute it to different sql instances.

**Review by jfunston — APPROVED** [review]

**Pull request merged**